nilearn/decomposition/dict_learning.py

```diff
@@ -0,0 +1,346 @@
+            the affine is considered the same for all.
+
+        confounds: CSV file path or 2D matrixf
+            This parameter is passed to nilearn.signal.clean. Please see the
+            related documentation for details
+        """
+        # Base logic for decomposition estimators
+        DecompositionEstimator.fit(self, imgs)
+
+        self.time_ = np.zeros(2)
+
+        if self.verbose:
+            print('[DictLearning] Loading data')
+        mask_reducer = MaskReducer(self.masker_,
+                                   reduction_ratio='auto',
+                                   n_components=self.n_components,
+                                   random_state=self.random_state,
+                                   memory_level=max(0, self.memory_level - 1),
+                                   n_jobs=self.n_jobs,
+                                   memory=self.memory,
+                                   temp_folder=self.temp_folder)
+        mask_reducer.fit(imgs, confounds)
+        data = mask_reducer.data_
+        if self.verbose:
+            print('[DictLearning] Learning initial components')
+        self._init_dict(data)
+
+        self._raw_fit(data)
+        mask_reducer.close()
+
+    def _raw_fit(self, data):
+        """Compute the mask and the maps across subjects, using raw_data. Can
+        only be called directly is dict_init and mask_img, or
+        components_init_ is provided
+
+        Parameters
+        ----------
+        data: ndarray,
+            Shape (n_samples, n_features)
+        """
+        if not hasattr(self, 'components_init_'):
+            if self.dict_init is not None:
+                if not hasattr(self, 'masker_'):
+                    DecompositionEstimator.fit(self, None)
+                self.components_init_ = self.masker_.transform(self.dict_init)
+            else:
+                raise ValueError('Calling _raw_fit directly is not possible '
+                                 'whithout providing dict_init and mask_img, '
+                                 'or setting components_init_')
+
+        if self.n_epochs < 0:
+            self.n_epochs = 1
+
+        n_samples, n_features = data.shape
+        if self.temp_folder is None:
+            in_core_batch_size = n_features
+        else:
+            in_core_batch_size = min(n_features, 1000)
+
+        batches = gen_batches(n_features, in_core_batch_size)
+        batches = chain.from_iterable(repeat(tuple(batches), self.n_epochs))
+        inner_stats = None
+        iter_offset = 0
+
+        if self.verbose:
+            print('[DictLearning] Computing initial loadings')
+        self._init_loadings(data)
+
+        dict_init = self._loadings_init
+        random_state = check_random_state(self.random_state)
+        stream_range = random_state.permutation(n_features)
+
+        if self.verbose:
+            print('[DictLearning] Learning dictionary')
+        for batch in batches:
```

Comment:
If we remove support for out of core via the memmap, batches are no longer critical here, and we use dict_learning_online with return_code=True.